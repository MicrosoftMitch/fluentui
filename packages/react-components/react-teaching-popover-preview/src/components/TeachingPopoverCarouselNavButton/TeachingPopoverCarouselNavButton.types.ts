import { PopoverContextValue } from '@fluentui/react-popover';
import type { ComponentState, ComponentProps, Slot } from '@fluentui/react-utilities';

export type TeachingPopoverCarouselNavButtonSlots = {
  /**
   * The root slot of the TeachingPopoverCarouselNavButton.
   * Default html element is div
   */
  root: NonNullable<Slot<'div'>>;
  /**
   * ARIA compliant nav buttons used to jump to pages
   */
  navButton: NonNullable<Slot<'button'>>;
};

/**
 * TeachingPopoverCarouselNavButton Props
 */
export type TeachingPopoverCarouselNavButtonProps = ComponentProps<Partial<TeachingPopoverCarouselNavButtonSlots>> & {
  /**
   * The page index that will be used to update carousel context
   */
  index: number;
};

/**
 * TeachingPopoverCarouselNavButton State
 */
export type TeachingPopoverCarouselNavButtonState = ComponentState<TeachingPopoverCarouselNavButtonSlots> & {
  /**
   * Enables selection state control
   */
  isSelected?: boolean;
} & Pick<PopoverContextValue, 'appearance'>;
