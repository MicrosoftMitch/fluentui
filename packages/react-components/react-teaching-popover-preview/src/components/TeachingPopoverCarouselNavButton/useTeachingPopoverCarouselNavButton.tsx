import { ARIAButtonSlotProps, useARIAButtonProps } from '@fluentui/react-aria';
import { usePopoverContext_unstable } from '@fluentui/react-popover';
import { useTabsterAttributes } from '@fluentui/react-tabster';
import { getIntrinsicElementProps, isHTMLElement, slot } from '@fluentui/react-utilities';
import * as React from 'react';

import type {
  TeachingPopoverCarouselNavButtonProps,
  TeachingPopoverCarouselNavButtonState,
} from './TeachingPopoverCarouselNavButton.types';
import { useCarouselContext_unstable } from '../TeachingPopoverCarousel/Carousel/CarouselContext';
import { useCarouselValues_unstable } from '../TeachingPopoverCarousel/Carousel/useCarouselValues';

/**
 * Create the state required to render TeachingPopoverCarouselNavButton.
 *
 * The returned state can be modified with hooks such as useTeachingPopoverCarouselNavButtonStyles_unstable,
 * before being passed to renderTeachingPopoverCarouselNavButton_unstable.
 *
 * @param props - props from this instance of TeachingPopoverCarouselNavButton
 * @param ref - reference to root HTMLElement of TeachingPopoverCarouselNavButton
 */
export const useTeachingPopoverCarouselNavButton_unstable = (
  props: TeachingPopoverCarouselNavButtonProps,
  ref: React.Ref<HTMLAnchorElement | HTMLButtonElement>,
): TeachingPopoverCarouselNavButtonState => {
  const { value, onClick, as = 'a' } = props;

  const appearance = usePopoverContext_unstable(context => context.appearance);
  const selectPageByValue = useCarouselContext_unstable(c => c.selectPageByValue);

  const values = useCarouselValues_unstable(values => values);
  let cValue = useCarouselContext_unstable(c => c.value);
  if (cValue === '') {
    cValue = values[0];
  }
  const index = values.indexOf(cValue ?? '');
  const totalPages = values.length;

  const isSelected = cValue === value;

  const setNewPage = React.useCallback(
    event => {
      onClick?.(event);

      if (!event.defaultPrevented && isHTMLElement(event.target)) {
        selectPageByValue(event, value);
      }
    },
    [onClick, value],
  );

  const defaultTabProps = useTabsterAttributes({
    focusable: { isDefault: isSelected },
  });

  const _carouselButton = slot.always<ARIAButtonSlotProps<'a'>>(
    getIntrinsicElementProps(as, useARIAButtonProps(props.as, props)),
    {
      elementType: 'button',
      defaultProps: {
        ref: ref as React.Ref<HTMLButtonElement & HTMLAnchorElement>,
        role: 'tab',
        type: 'button',
        'aria-label': `${index + 1} of ${totalPages}`,
        ...defaultTabProps,
      },
    },
  );

  _carouselButton.onClick = setNewPage;

  const state: TeachingPopoverCarouselNavButtonState = {
    isSelected,
    appearance,
    components: {
      root: 'button',
    },
    root: _carouselButton,
  };

  return state;
};
