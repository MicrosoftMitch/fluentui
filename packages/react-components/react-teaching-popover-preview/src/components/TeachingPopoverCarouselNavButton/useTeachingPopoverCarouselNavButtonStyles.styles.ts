import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
import type { SlotClassNames } from '@fluentui/react-utilities';
import type {
  TeachingPopoverCarouselNavButtonSlots,
  TeachingPopoverCarouselNavButtonState,
} from './TeachingPopoverCarouselNavButton.types';
import { tokens } from '@fluentui/react-theme';
import { createCustomFocusIndicatorStyle } from '@fluentui/react-tabster';

export const teachingPopoverCarouselNavButtonClassNames: SlotClassNames<TeachingPopoverCarouselNavButtonSlots> = {
  root: 'fui-TeachingPopoverCarouselNavButton',
  navButton: 'fui-TeachingPopoverCarouselNavButton__navButton',
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    display: 'flex',
    ...createCustomFocusIndicatorStyle(
      {
        ...shorthands.outline('2px', 'solid', tokens.colorStrokeFocus2),
        ...shorthands.borderRadius(tokens.borderRadiusCircular),
      },
      { selector: 'focus-within' },
    ),
  },
  rootSelected: {
    display: 'flex',
    ...createCustomFocusIndicatorStyle(
      {
        ...shorthands.outline('2px', 'solid', tokens.colorStrokeFocus2),
        ...shorthands.borderRadius('4px'),
      },
      { selector: 'focus-within' },
    ),
  },
  navButton: {
    cursor: 'pointer',
    boxSizing: 'border-box',
    backgroundColor: tokens.colorBrandBackground,
    ...shorthands.border(0),
    ...shorthands.borderRadius('50%'),
    ...shorthands.padding('0px'),
  },
  navButtonSelected: {
    height: '8px',
    width: '16px',
    ...shorthands.border(0),
    ...shorthands.borderRadius('4px'),
    ...shorthands.padding('0px'),
  },
  navButtonUnselected: {
    height: '8px',
    width: '8px',
    opacity: 0.3,
  },
  navButtonBrand: {
    backgroundColor: tokens.colorNeutralForegroundOnBrand,
  },
});

/**
 * Apply styling to the TeachingPopoverCarouselNavButton slots based on the state
 */
export const useTeachingPopoverCarouselNavButtonStyles_unstable = (
  state: TeachingPopoverCarouselNavButtonState,
): TeachingPopoverCarouselNavButtonState => {
  const styles = useStyles();
  const { appearance, isSelected } = state;

  state.root.className = mergeClasses(
    teachingPopoverCarouselNavButtonClassNames.root,
    isSelected ? styles.rootSelected : styles.root,
    state.root.className,
  );

  if (state.navButton) {
    const selectedButtonStyle = isSelected ? styles.navButtonSelected : styles.navButtonUnselected;
    state.navButton.className = mergeClasses(
      teachingPopoverCarouselNavButtonClassNames.navButton,
      styles.navButton,
      selectedButtonStyle,
      appearance === 'brand' ? styles.navButtonBrand : undefined,
      state.navButton.className,
    );
  }

  return state;
};
