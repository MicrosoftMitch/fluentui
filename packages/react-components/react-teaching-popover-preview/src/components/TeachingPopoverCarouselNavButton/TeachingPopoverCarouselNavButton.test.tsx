import * as React from 'react';
import { render } from '@testing-library/react';
import { TeachingPopoverCarouselNavButton } from './TeachingPopoverCarouselNavButton';

describe('TeachingPopoverCarouselNavButton', () => {
  it('renders a default state', () => {
    const result = render(<TeachingPopoverCarouselNavButton value={'test-1'} />);
    expect(result.container).toMatchSnapshot();
  });
});
