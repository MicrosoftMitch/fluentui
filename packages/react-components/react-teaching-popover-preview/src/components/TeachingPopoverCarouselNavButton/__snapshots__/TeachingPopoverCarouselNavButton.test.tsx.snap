// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeachingPopoverCarouselNavButton renders a default state 1`] = `
<div>
  <button
    aria-label="0 of 0"
    class="fui-TeachingPopoverCarouselNavButton"
    data-tabster="{\\"focusable\\":{\\"isDefault\\":false}}"
    role="tab"
    type="button"
  />
</div>
`;
